PowerEditor/src/WinControls/ProjectPanel/Directory.h

```diff
@@ -0,0 +1,126 @@
+	Directory(bool hideEmptyDirs = true);
+	// directory with path - is instantly read in ctor, if autoread==true
+	Directory(const generic_string& path, const std::vector<generic_string>& filters = std::vector<generic_string>(), bool hideEmptyDirs=true, bool autoread=true );
+
+	virtual ~Directory() {}
+
+	const generic_string& getPath() const { return _path; }
+
+	// read the directory. Returns true, if the directory exists.
+	bool read(const generic_string& path, const std::vector<generic_string>& filters);
+	bool read(const generic_string& path) { return read(path,_filters); }
+	bool read() { return read(_path, _filters); }
+
+	// re-read the current directory, but only if it was changed.
+	// returns true, if directory was re-read, false if directory has not changed.
+	bool readIfChanged(bool respectEmptyDirs = true);
+
+	// false when not initialized.
+	bool exists() const { return _exists; }
+
+	const std::set<generic_string,comparator>& getDirs() const { return _dirs; }
+	const std::set<generic_string,comparator>& getFiles() const { return _files; }
+
+	bool operator== (const Directory& other) const;
+	bool operator!= (const Directory& other) const;
+
+	bool empty() const { return _dirs.empty() && _files.empty(); }
+
+	bool writeTimeHasChanged() const;
+
+	void setFilters(const std::vector<generic_string>& filters = std::vector<generic_string>(), bool autoread=true);
+	const std::vector<generic_string>& getFilters() const { return _filters; }
+
+	// empty directories can be hidden:
+	// "empty" means in this case, neither the directory itself nor its subdirectories contains any data, which match the filters
+	bool getHideEmptyDirs() const { return _hideEmptyDirs; }
+	void setHideEmptyDirs(bool hideEmptyDirs, bool autoread=true);
+
+	// synchronizeTo is basically like a copy constructor, with the difference, that it calls the
+	// following virtual functions onDirAdded(), ...
+	// Using this makes only sense, if you have subclassed this and evaluate those virtual methods; otherwise just use the copy ctor/assigment operator.
+	// Note that the virtual methods are called BEFORE the contents of this directory are really changed.
+	void synchronizeTo(const Directory& other);
+
+protected:
+
+	virtual void onBeginSynchronize(const Directory&) {}
+	virtual void onDirAdded(const generic_string&) {}
+	virtual void onDirRemoved(const generic_string&) {}
+	virtual void onFileAdded(const generic_string&) {}
+	virtual void onFileRemoved(const generic_string&) {}
+	virtual void onEndSynchronize(const Directory&) {}
+
+private:
+
+	static void enablePrivileges();
+	static bool enablePrivilege(LPCTSTR privName);
+	void append(const generic_string& path, const generic_string& filter, bool readDirs);
+	bool containsDataChanged() const;
+	bool containsData(const generic_string& path) const;
+	bool containsData(const generic_string& path, const generic_string& filter) const;
+	bool readLastWriteTime(FILETIME& filetime) const;
+
+
+};
+
+
+#endif
```

Comment:
Grr.
![directory h_endl](https://cloud.githubusercontent.com/assets/2142308/8297005/dff55fbe-1925-11e5-8ece-1d1de4348a3e.png)

A new line at the end of the file is always a good idea.